### PR TITLE
IBM API to submit job and retrieve result

### LINKED ISSUE #13: Research IBM API
Steps to start migrating to use IBM API 

- [x] Document steps to acquire an access token
- [x] Figure out what IBM API endpoint to call to submit a job
- [x] Try a sample script to submit a job and document response (try sending the json we have from poc/client.py
- [x] Figure out IBM API endpoint to retrieve results
- [ ] Sample request to retrieve the result and check its format

Report back if we need any changes to the input format and other issues that we might potentially face while using the IBM backend.

### PR DESCRIPTION
**Pull request by srihitha97**
To pass the circuit to the backend, we must first map it onto the backend, package it, and send to the device. This is all done by using the execute function:

`job = execute(circuit, backend)`

Alternatively, the circuit can be mapped using the transpile function, package it using assemble, and then send it from the backend instance itself.

Result on running ibm_api.py:

```

[<AccountProvider for IBMQ(hub='ibm-q', group='open', project='main')>]


The backends are:
[<IBMQSimulator('ibmq_qasm_simulator') from IBMQ(hub='ibm-q', group='open', project='main')>, <IBMQBackend('ibmqx2') from IBMQ(hub='ibm-q', group='open', p
roject='main')>, <IBMQBackend('ibmq_16_melbourne') from IBMQ(hub='ibm-q', group='open', project='main')>, <IBMQBackend('ibmq_armonk') from IBMQ(hub='ibm-q'
, group='open', project='main')>, <IBMQBackend('ibmq_athens') from IBMQ(hub='ibm-q', group='open', project='main')>, <IBMQBackend('ibmq_santiago') from IBM
Q(hub='ibm-q', group='open', project='main')>, <IBMQBackend('ibmq_lima') from IBMQ(hub='ibm-q', group='open', project='main')>, <IBMQBackend('ibmq_belem')
from IBMQ(hub='ibm-q', group='open', project='main')>, <IBMQBackend('ibmq_quito') from IBMQ(hub='ibm-q', group='open', project='main')>, <IBMQSimulator('si
mulator_statevector') from IBMQ(hub='ibm-q', group='open', project='main')>, <IBMQSimulator('simulator_mps') from IBMQ(hub='ibm-q', group='open', project='
main')>, <IBMQSimulator('simulator_extended_stabilizer') from IBMQ(hub='ibm-q', group='open', project='main')>, <IBMQSimulator('simulator_stabilizer') from
 IBMQ(hub='ibm-q', group='open', project='main')>]
 
 
Selecting backend ibmq_qasm_simulator...
ibmq_qasm_simulator


Printing the status of the job
JobStatus.VALIDATING


The job ID of the job is:
6053893a4f7d2b415a96bcfb


The job result is:
Result(backend_name='ibmq_qasm_simulator', backend_version='0.7.6', qobj_id='6448e41c-b8eb-482c-a133-a31e77c78ecd', job_id='60538802ec3b4031b4a8029b', succ
ess=True, results=[ExperimentResult(shots=1024, success=True, meas_level=MeasLevel.CLASSIFIED, data=ExperimentResultData(counts={'0x0': 520, '0x3': 504}),
header=QobjExperimentHeader(clbit_labels=[['c', 0], ['c', 1]], creg_sizes=[['c', 2]], global_phase=0.0, memory_slots=2, n_qubits=2, name='circuit7', qreg_s
izes=[['q', 2]], qubit_labels=[['q', 0], ['q', 1]]), status=DONE, metadata={'fusion': {'enabled': False}, 'measure_sampling': True, 'method': 'stabilizer',
 'parallel_shots': 1, 'parallel_state_update': 16}, seed_simulator=166211994, time_taken=0.0015332310000000001)], date=2021-03-18 13:04:05.721994-04:00, st
atus=COMPLETED, status=QobjHeader(backend_name='ibmq_qasm_simulator', backend_version='0.1.547'), metadata={'max_memory_mb': 515872, 'omp_enabled': True, '
parallel_experiments': 1, 'time_taken': 0.0015976150000000001}, time_taken=0.0015976150000000001, client_version={'qiskit': '0.23.5'})
{'00': 520, '11': 504}
```

### CODE REVIEW

File: poc/ibm_api.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from qiskit import *
+from qiskit import IBMQ
+from client import *
+
+IBMQ.save_account('507a5316e425f15330cfdaa0e7fc473429105af797d5b11612a4ef511ca37af70ec3116afde55fad56352b92a9b7636c921b1b35c0ee7334f65bba27f7d1e692')
```
Comment (rsquare2010) [inline poc/ibm_api.py]: Open-source project, not a good idea to have the token in pushed code.

Comment (saisankargochhayat) [reply poc/ibm_api.py]: Might wanna delete this and recreate a key as well.

File: poc/ibm_api.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+print(job.status())
+
+backend_temp = job.backend()
+backend_temp
+
+print("The job ID of the job is:")
+job.job_id()
+
+#Getting the job result
+
```
Comment (saisankargochhayat) [inline poc/ibm_api.py:60]: So we need access to the job result just with the job-id without using the old job object,

Comment (srihitha97) [reply poc/ibm_api.py:60]: Yes, now getting job from job id and getting result of that

**Review by rsquare2010 — APPROVED** [review]

**Review by saisankargochhayat — APPROVED** [review]
LGTM!

**Comment by saisankargochhayat** [comment]
@srihitha97 if you are planning to move the get result part into a diff file we can merge this after that?

File: poc/ibm_api.py (hunk 1 above)
Comment (husky-parul) [reply poc/ibm_api.py]: The token will be read from secret file mounted on the pod.

File: poc/ibm_api.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+print("The backends are:")
+print(provider.backends())
+
+print("Selecting backend ibmq_qasm_simulator...")
+backend = provider.get_backend('ibmq_qasm_simulator')
+print(backend)
+
+#Submitting a job
+
+job = execute(qc, backend)
```
Comment (husky-parul) [inline poc/ibm_api.py]: I would refrain from using the `execute`. From what I remember execute is not asynchronous and the client calling execute waits for the job to be added on the backend queue. Use `transpile` and then `assemble` to get the JSON schema of the object.

```
qc = QuantumCircuit(2, 2)
qc.h(0)
qc.cx(0, 1)
qc.measure([0, 1], [0, 1])
backend_sim = BasicAer.get_backend('qasm_simulator')
transpiled_qc = transpile(qc, backend_sim)
qasm_obj = assemble(transpiled_qc)
```

Comment (husky-parul) [reply poc/ibm_api.py]: Follow this https://github.com/qiskit-community/openshift-quantum-operators/blob/master/operators-examples/openshift-ibm-quantum-operator/getting_started.md

Comment (srihitha97) [reply poc/ibm_api.py]: Sure @husky-parul , will make that change

**Comment by srihitha97** [comment]
> 
> 
> @srihitha97 if you are planning to move the get result part into a diff file we can merge this after that?

Yes, will make the above 2 changes and request for review again.

**Review by saisankargochhayat — APPROVED** [review]
LGTM!

**Pull request merged**